src/awkward/_connect/pyarrow.py

```diff
@@ -282,6 +282,11 @@ def mask_parameters(awkwardarrow_type):
         return None
 
 
+def revertable(modified, original):
+    modified.__pyarrow_original = original
+    return modified
```

Comment:
The first constraint (in BitMaskedArray) revealed that the from-pyarrow code was assuming that we could temporarily wrap something with an option-type array and then remove it later.

We actually can't because the outer layer, with `simplified`, could merge in some non-revertable way with its contents.

To fix that, we now hold onto both the (possibly `simplified`) option-type array and the original array, so that reverting it is a matter of just pulling out the original, held as a `__pyarrow_original` attribute. (Don't you love Python?)

`ak.from_arrow` is a gatekeeper for `handle_pyarrow`/`popbuffers`/`popbuffers_finalize`. It's a private, internal implementation detail (notwithstanding the fact that some tests call it directly). All users, including third-party developers, must access it through `ak.from_arrow`, and this function now removes the `__pyarrow_original` attribute.